cache/image_cache.go

```diff
@@ -0,0 +1,82 @@
+package cache
+
+import (
+	"encoding/json"
+	"io"
+
+	"github.com/pkg/errors"
+
+	"github.com/buildpack/lifecycle/image"
+	"github.com/buildpack/lifecycle/metadata"
+)
+
+//go:generate mockgen -package testmock -destination testmock/image_factory.go github.com/buildpack/lifecycle/cache ImageFactory
+type ImageFactory interface {
+	NewEmptyLocal(string) image.Image
+}
+
+type ImageCache struct {
+	factory   ImageFactory
+	origImage image.Image
+	newImage  image.Image
+}
+
+func NewImageCache(factory ImageFactory, origImage image.Image) *ImageCache {
+	newImage := factory.NewEmptyLocal(origImage.Name())
+	return &ImageCache{
+		factory:   factory,
+		origImage: origImage,
+		newImage:  newImage,
+	}
+}
+
+func (c *ImageCache) Name() string {
+	return c.origImage.Name()
+}
+
+func (c *ImageCache) SetMetadata(metadata Metadata) error {
+	data, err := json.Marshal(metadata)
+	if err != nil {
+		return errors.Wrap(err, "serializing metadata")
+	}
+	return c.newImage.SetLabel(MetadataLabel, string(data))
+}
+
+func (c *ImageCache) RetrieveMetadata() (Metadata, error) {
+	contents, err := metadata.GetRawMetadata(c.origImage, MetadataLabel)
+	if err != nil {
+		return Metadata{}, errors.Wrap(err, "retrieving metadata")
+	}
+
+	meta := Metadata{}
+	_ = json.Unmarshal([]byte(contents), &meta)
```

Comment:
Don't ignore this error:
```
return meta, json.Unmarshal([]byte(contents), &meta)
```

Reply:
@sclevine This is intentional for the cache behavior. In this case, if the metadata is malformed, we're returning an empty metadata object.

Reply:
if `json.Unmarshal` returns an error, you have no guarantee that the values returned are safe to use. I suggest checking for an error and returning empty struct (with no error) if you'd like to accept malformed metadata.

Reply:
Gotcha. I didn't realize the same thing goes for `meta` since it's not technically returned. But I see your point and think it's clearer to do what you mentioned.

Reply:
Approving conditional on this change 👍 

For some more context: `json.Unmarshal` acts directly on the fields in `meta` (without copying it first or rolling back on failure), and doesn't reset anything when it encounters an error (it actually panics and runs up the stack).

Reply:
Great point @sclevine. Thanks for the context. Done.